arrows/klv/klv_read_write.txx

```diff
@@ -343,6 +356,114 @@ klv_ber_oid_length( T value )
   return ( _int_bit_length( value ) + 6 ) / 7;
 }
 
+// ---------------------------------------------------------------------------
+template < class T, class Iterator >
+double
+klv_read_flint( double minimum, double maximum, Iterator& data, size_t length )
+{
+  KLV_ASSERT_INT( T );
+  KLV_ASSERT_UINT8_ITERATOR( data );
+
+  _check_range_length( minimum, maximum, length );
+
+  // Before read to avoid moving iterator
+  if( std::is_signed< T >::value && minimum != -maximum )
+  {
+    throw std::logic_error( "range must be symmetrical around zero" );
+  }
+
+  auto const int_value = klv_read_int< T >( data, length );
+  auto const float_value = static_cast< double >( int_value );
+
+  // C++17: if constexpr
+  if( std::is_signed< T >::value )
+  {
+    // Special invalid / out-of-range value
+    if( int_value == _int_min< T >( length ) )
+    {
+      return std::numeric_limits< double >::quiet_NaN();
+    }
+
+    auto const scale = maximum / _int_max< T >( length );
+    return float_value * scale;
+  }
+  else
+  {
+    auto const scale = ( maximum - minimum ) / _int_max< T >( length );
+    return float_value * scale + minimum;
+  }
+}
+
+// ---------------------------------------------------------------------------
+template < class T, class Iterator >
+void
+klv_write_flint( double value, double minimum, double maximum, Iterator& data,
+                 size_t length )
+{
+  // Ensure types are compatible with our assumptions
+  KLV_ASSERT_INT( T );
+  KLV_ASSERT_UINT8_ITERATOR( data );
+
+  _check_range_length( minimum, maximum, length );
+
+  // C++17: if constexpr
+  if( std::is_signed< T >::value )
+  {
+    if( minimum != -maximum )
+    {
+      throw std::logic_error( "range must be symmetrical around zero" );
+    }
+
+    auto const max_int = _int_max< T >( length );
+    auto const min_int = _int_min< T >( length );
+    auto const scale = max_int / maximum;
+    auto const float_value = value * scale;
+    auto int_value = static_cast< T >( std::round( float_value ) );
+
+    if( !( value <= maximum && minimum <= value ) )
```

Comment:
I was going to (weakly) suggest:
```suggestion
    if( !( minimum <= value && value <= maximum ) )
```

...however, am I reading this correctly that this will catch *any* out-of-range value? Doesn't that mean the next two branches are unreachable? (I guess they should be moved inside this block; I like that this will also catch NaN... which `if( value < minimum || value > maximum )` wouldn't, and is probably why you wrote it this way.)

Maybe I'm confused?

Reply:
I found out that these were necessary while writing the unit tests; you correctly pointed out the possibility of weird behavior above. It is possible that, due to rounding error, `value` can be between `minimum` and `maximum`, but  `value * scale` (which is supposed to vary between `min_int` and `max_int`) is greater than `max_int`. For sufficiently large `scale` (i.e. sufficiently many bytes in the integer), these rounding errors can be greater than 1. This causes issues with integer overflow when casting to `int_value`. Hence, the extra checks.

Reply:
>If I am reading this correctly, any out-of-range value will be mapped to the special "NaN" value, not clamped to the min/max. Is this intended?

Yes. The [standard](https://gwg.nga.mil/misb/docs/standards/ST0601.17.pdf) says at e.g. Section 8.6: `0x8000 = "Out of Range" indicator`. `NaN` is also mapped to that special value because it presumably counts as out of range. When reading in, that value is mapped to `NaN` because what else would we map it to. So even though in the code it is treated as a `NaN` equivalent, in fact its primary purpose is as an "out of range" indicator.

> The documentation says "out-of-range values are clamped".

For unsigned values only; I can reword this if it is unclear.